### ISSUE TITLE
Add more contrast for `document_highlight()`

### ISSUE THREAD

**Issue by strash** [ramojus/mellifluous.nvim#39]
On the light scheme the highlight is fine but on the dark scheme it is almost invisible.
It would be nice to add some contrast, just enough to be distinguishable but not annoying.
  
<img width="676" alt="Frame 2242" src="https://github.com/ramojus/mellifluous.nvim/assets/14233263/c57cead7-4aaf-4197-a9c9-3b90847f3ff6">

<img width="627" alt="Screenshot 2024-06-05 at 23 08 15" src="https://github.com/ramojus/mellifluous.nvim/assets/14233263/e60d4241-5547-466f-96a5-b89ac27bda8e">

<img width="779" alt="Screenshot 2024-06-05 at 23 09 41" src="https://github.com/ramojus/mellifluous.nvim/assets/14233263/a57b8062-36a8-4a9d-b8d1-79675d45cc2e">

**Comment by ramojus** [comment]
I think these could be a bit lighter on dark bg. On light bg they do seem to be more visible.
I may think about configuration here, but I am currently not convinced that it's worthwhile. Let's see when this change is applied.

I am currently refactoring and changing how shades are applied everywhere, so, as part of that, I will try to make these a bit lighter while keeping the low contrast vibe.

**Comment by antoineco** [comment]
One of the difficulties here is to not overlap with the `Search` highlight group.

In the screenshot below, `dict` is highlighted as the result of a current search (the background is slightly lighter), whereas `msg` is highlighted by `vim.lsp.buf.document_highlight()`:
 
![image](https://github.com/ramojus/mellifluous.nvim/assets/3299086/510d37d9-38df-48d5-98a3-5c4720022079)

Using a colorful background for searches would solve that and "free" `bg4` for the `LspReferenceText` highlight group. Here is how Zenbones does it, for comparison:

![image](https://github.com/ramojus/mellifluous.nvim/assets/3299086/75eb8e58-8048-416e-adea-7247b88092a3)

**Comment by antoineco** [comment]
By the way, out of interest I compared the APCA Lc[^1] value for `LspReferenceText` in both dark and light modes. @strash is right, the contrast is objectively lower in dark mode (9 vs. 0).

Here are the values:

- bg_contrast = 'soft'
  - dark: Lc: [0.0](https://www.myndex.com/APCA/?BG=1f1f1f&TXT=2d2d2d&DEV=G4g&BUF=A22)
  - light: Lc [-9.0](https://www.myndex.com/APCA/?BG=e1e1e1&TXT=f2f2f2&DEV=G4g&BUF=A22)
- bg_contrast = 'medium'
  - dark: Lc: [0.0](https://www.myndex.com/APCA/?BG=1a1a1a&TXT=282828&DEV=G4g&BUF=A22)
  - light: Lc [-9.0](https://www.myndex.com/APCA/?BG=1a1a1a&TXT=282828&DEV=G4g&BUF=A22)

APCA is the algorithm used by [W3C](https://www.w3.org/) in the [draft version](https://github.com/w3c/wcag/issues/695) of the WCAG accessibility guidelines. It is a very useful tool for designing colorschemes and ensuring that contrasts are proportional between dark and light modes.

[^1]:  Lightness contrast

**Comment by ramojus** [comment]
Interesting. My eyes don't agree with this algorithm though 😀 Like if I try to make dark bg contrast match the contrast of light bg, it doesn't look like the same contrast at all...

![image](https://github.com/ramojus/mellifluous.nvim/assets/41536253/80eca77e-01f9-4b61-97c5-438aadb0f565)
![image](https://github.com/ramojus/mellifluous.nvim/assets/41536253/d4f7eab9-5a8a-4d06-a693-5c01f1e02c44)

I guess it also depends a lot on the environment and the display, e.g. I get much higher contrast on my laptop screen than on my monitor. And normally you would (at least I would) use light bg in bright environment compared to dark bg in darker environment. So even if light bg objectively does have more contrast, it would make sense when considering the practicality of described scenario.

**Comment by ramojus** [comment]
@strash would you say [this](https://github.com/ramojus/mellifluous.nvim/pull/40) is enough of a difference? I'm hesitant to add more contrast, as it will probably start being distracting. I would say it's still lower contrast than with light bg, but as I said before, dark bg is more targeted towards not-that-bright environments.

**Comment by antoineco** [comment]
> My eyes don't agree with this algorithm [...]
> even if light bg objectively does have more contrast, it would make sense when considering the practicality of described scenario.

That's fair, and to be clear I don't disagree with the existing highlights, and use primarily dark mode.
Keep in mind that although the algorithm takes into accounts the different needs of light and dark modes, it is geared towards _accessibility use cases_, primarily for text/background ratios.

**Comment by strash** [comment]
> @strash would you say [this](https://github.com/ramojus/mellifluous.nvim/pull/40) is enough of a difference? I'm hesitant to add more contrast, as it will probably start being distracting. I would say it's still lower contrast than with light bg, but as I said before, dark bg is more targeted towards not-that-bright environments.

@ramojus It's hard to say. I see the difference between "before" and "after" but I also see a big difference between your screenshots and mine, so I'm a bit confused. But overall it's better now.

<img width="1401" alt="Group 2242" src="https://github.com/ramojus/mellifluous.nvim/assets/14233263/3b61e6d9-7772-4e40-80ab-df46f211c4f9">

**Comment by ramojus** [comment]
@strash Probably not what you meant, but I'll say this just in case - I wasn't trying to match your suggestion (on your second screenshot) exactly. Instead, I was aiming for a more subtle change.

I think the difference between my before screenshot and your unmodified screenshot is that the font you are using has thicker strokes compared to the font that I'm using. That makes all of the colors appear brighter (as they have more area to fully show themselves, without antialiasing "weakening" them), and thus the contrast of bg layers is harder to see. I've checked with a color picker, and `LspReferenceText` bg is the same color in both screenshots.

I think different fonts affecting the perceived contrast of bg layers may be enough of a reason for a config option. I'll think about it.

**Comment by strash** [comment]
@ramojus should I close the issue?

**Comment by ramojus** [comment]
Better to keep it open for now. I'll add configuration for bg layers contrast and that's a bit related.

**Comment by ramojus** [comment]
@strash I've added the ability to override shades in #45, but it's a breaking change, so it was merged to v1 branch. After prolonged usage, I ended not liking the changes in #40, so I will revert them. This is how you can replicate them (from v1 branch):
```lua
require 'mellifluous'.setup({
    color_overrides = {
        dark = {
            colors = function(colors)
                return {
                    bg2 = colors.bg:lightened(4),
                    bg3 = colors.bg:lightened(7),
                    bg4 = colors.bg:lightened(10),
                    bg5 = colors.bg:lightened(13),
                }
            end,
        }
    },
})
```
`LspReferenceText` is linked to bg3 -- you can change it if you want even more contrast.

After a couple of weeks, v1 branch should be merged to main (when v1.0.0 will be released).
